### PR TITLE
[MAINTENANCE] `datasource`, `data_context`, `core` typing, lint fixes

### PR DESCRIPTION
**Pull request by Kilo59**
Changes proposed in this pull request:
- linting and typing for various modules
  - `data_context`
  - `datasource` 
  - low error count `/core` modules
- revert some config changes
- add `--report` flag to `invoke type-check`

## Notes

* There are only a couple of actual run-time changes and I've called them out with comments for extra scrutiny.
* 90% of the "fixes" are in-fact just warning supression comments so that we can start type-checking these modules and stop accumulating typing debt.

## Type Coverage reports
Can now pass a `--report` flag to `invoke type-check` in order to generate type coverage reports.

image

image

### CODE REVIEW

File: great_expectations/profile/basic_dataset_profiler.py [hunk 1]
```diff
@@ -123,7 +123,7 @@ class BasicDatasetProfiler(BasicDatasetProfilerBase):
     such as min, max, mean and median, for numeric columns, and distribution of values, when appropriate.
     """
 
-    @classmethod
+    @classmethod  # noqa: C901
```
Comment (Kilo59) [inline great_expectations/profile/basic_dataset_profiler.py:126]: This is pretty annoying.
flake8 needs `noqa` statements on both the method and `@classmethod` decorator.

File: pyproject.toml [hunk 2]
```diff
@@ -32,24 +32,15 @@ exclude = [
...
     'core/expectation_validation_result\.py', # 9
-    'core/run_identifier\.py', # 1
-    'core/yaml_handler\.py', # 1
-    'datasource', # 93
+    'datasource/batch_kwargs_generator', # 16
+    'datasource/data_connector', # 60
     'dataset',  # 19
-    'data_context/data_context/abstract_data_context.py', # 84
     'data_context/data_context/base_data_context.py', # 76
-    'data_context/data_context/cloud_data_context.py', # 17
```
Comment (Kilo59) [inline pyproject.toml]: For a summary of the coverage changes, see above ☝️

File: tasks.py [hunk 3]
```diff
@@ -117,7 +117,9 @@ def type_coverage(ctx):
     try:
         check_type_hint_coverage.main()
     except AssertionError as err:
-        raise invoke.Exit(message=str(err), code=1)
+        raise invoke.Exit(
+            message=f"{err}\n\n  See {check_type_hint_coverage.__file__}", code=1
+        )
```
Comment (Kilo59) [inline tasks.py:122]: Adding a link to the `check_type_hint_coverage` file when this fails

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 4]
```diff
@@ -1158,7 +1160,7 @@ def _get_global_config_value(
         assert (conf_file_section and conf_file_option) or (
             not conf_file_section and not conf_file_option
         ), "Must pass both 'conf_file_section' and 'conf_file_option' or neither."
-        if environment_variable and os.environ.get(environment_variable, False):
+        if environment_variable and os.environ.get(environment_variable, ""):
```
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py:1163]: Here's an actual run-time change.
`os.environ.get()` wants a `str` default value.

File: great_expectations/datasource/data_connector/inferred_asset_sql_data_connector.py [hunk 5]
```diff
@@ -166,13 +166,13 @@ def _refresh_introspected_assets_cache(
...
+            if splitter_method is not None:
                 data_asset_config["splitter_method"] = splitter_method
-            if not splitter_kwargs is None:
+            if splitter_kwargs is not None:
                 data_asset_config["splitter_kwargs"] = splitter_kwargs
-            if not sampling_method is None:
+            if sampling_method is not None:
                 data_asset_config["sampling_method"] = sampling_method
-            if not sampling_kwargs is None:
+            if sampling_kwargs is not None:
```
Comment (Kilo59) [inline great_expectations/datasource/data_connector/inferred_asset_sql_data_connector.py:175]: Here's another case of actual run-time changes.
As suggested by flake8

https://www.flake8rules.com/rules/E714.html

File: great_expectations/datasource/sqlalchemy_datasource.py [hunk 6]
```diff
@@ -33,7 +33,7 @@
     logger.debug("Unable to import sqlalchemy.")
 
 
-if sqlalchemy != None:
+if sqlalchemy is not None:
```
Comment (Kilo59) [inline great_expectations/datasource/sqlalchemy_datasource.py:36]: Run-time change

**Comment by cdkini** [comment]
Disabling auto-merge in order to prioritize the 0.15.20 release - will review and set back upon completion of release process.

File: great_expectations/core/__init__.py [hunk 7]
```diff
@@ -21,22 +21,22 @@
...
+    "ExpectationSuiteValidationResultSchema",
+    "ExpectationValidationResult",
+    "ExpectationValidationResultSchema",
+    "expectationSuiteValidationResultSchema",
+    "expectationValidationResultSchema",
+    "get_metric_kwargs_id",
+    "IDDict",
+    "RunIdentifier",
+    "RunIdentifierSchema",
+    "ge_urn",
```
Comment (cdkini) [inline great_expectations/core/__init__.py:39]: Thoughts on doing `obj.__class__.__name__`? What if we rename any of these items?

File: great_expectations/datasource/data_connector/inferred_asset_sql_data_connector.py (hunk 5 above)
Comment (cdkini) [reply great_expectations/datasource/data_connector/inferred_asset_sql_data_connector.py:175]: love it!

**Review by cdkini — APPROVED** [review]

**Pull request merged**